src/test/java/com/fasterxml/jackson/databind/jsontype/TestPolymorphicDeductionObjectVsArray.java

```diff
@@ -0,0 +1,103 @@
+package com.fasterxml.jackson.databind.jsontype;
+
+import com.fasterxml.jackson.annotation.*;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.testutil.DatabindTestUtil;
+import org.junit.jupiter.api.Test;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+public class TestPolymorphicDeductionObjectVsArray extends DatabindTestUtil {
+    @JsonTypeInfo(use = JsonTypeInfo.Id.DEDUCTION, defaultImpl = DataArray.class)
+    @JsonSubTypes({@JsonSubTypes.Type(DataObject.class), @JsonSubTypes.Type(DataArray.class)})
+    interface Data {
+        @JsonIgnore
+        boolean isObject();
+    }
+
+    static class DataItem {
+        final String id;
+
+        @JsonCreator(mode = JsonCreator.Mode.PROPERTIES)
+        DataItem(@JsonProperty("id") String id) {
+            this.id = id;
+        }
+
+        public String getId() {
+            return id;
+        }
+    }
+
+    static class DataObject extends DataItem implements Data {
+
+        @JsonCreator(mode = JsonCreator.Mode.PROPERTIES)
+        DataObject(@JsonProperty("id") String id) {
+            super(id);
+        }
+
+        @Override
+        public boolean isObject() {
+            return true;
+        }
+    }
+
+    static class DataArray extends ArrayList<DataItem> implements Data {
+        @JsonCreator
+        DataArray(Collection<DataItem> items) {
+            super(new ArrayList<>(items));
+        }
+
+        @Override
+        public boolean isObject() {
+            return false;
+        }
+    }
+
+    static class Container {
+        @JsonProperty("data")
+        Data data;
+    }
+
+    private final ObjectMapper MAPPER = new ObjectMapper();
+
+    private static final String containerWithObjectData = a2q("{'data':{'id':'#1'}}");
+
+    private static final String containerWithArrayData = a2q("{'data':[{'id':'#1'}]}");
+
+    @Test
+    public void testDeserialization() throws Exception {
+        Container container = MAPPER.readValue(containerWithObjectData, Container.class);
+
+        assertInstanceOf(DataObject.class, container.data);
+        assertSame(container.data.getClass(), DataObject.class);
+        assertTrue(container.data.isObject());
+        assertEquals("#1", ((DataItem) container.data).id);
+
+        container = MAPPER.readValue(containerWithArrayData, Container.class);
+        assertInstanceOf(DataArray.class, container.data);
+        assertSame(container.data.getClass(), DataArray.class);
+        assertFalse(container.data.isObject());
+
+        var arrayDataIterator = ((Iterable<DataItem>) container.data).iterator();
```

Comment:
No `var`, we are still Java 8 (I can change)

Reply:
Oops, so used to it, my bad. Thanks for the change.